File: removeNthNodeFromEndOfList.py
# problem: https://leetcode.com/problems/remove-nth-node-from-end-of-list/
# Runtime: 28 ms, faster than 91.78% of Python3 online submissions for Remove Nth Node From End of List.
# Memory Usage: 14.3 MB, less than 48.74% of Python3 online submissions for Remove Nth Node From End of List.


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        walker = head
        runner = head
        for _ in range(n):
            runner = runner.next
        prev = None
        while runner:
            prev = walker
            walker = walker.next
            runner = runner.next
        if prev:
            prev.next = walker.next
        elif walker == head:
            head = walker.next
        return head

File: removeNthNodeFromEndOfList.cpp
// problem: https://leetcode.com/problems/remove-nth-node-from-end-of-list/
// Runtime: 10 ms, faster than 33.90% of C++ online submissions for Remove Nth
// Node From End of List. Memory Usage: 10.7 MB, less than 75.00% of C++ online
// submissions for Remove Nth Node From End of List.

#include <bits/stdc++.h>

using namespace std;

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  ListNode *removeNthFromEnd(ListNode *head, int n) {
    ListNode *walker = head;
    ListNode *runner = head;
    while (n--)
      runner = runner->next;
    if (!runner)
      return head->next;
    while (runner && runner->next) {
      walker = walker->next;
      runner = runner->next;
    }
    ListNode *next = walker->next;
    walker->next = next != nullptr ? next->next : nullptr;
    return head;
  }
};
